packages/moments/src/client/MomentsClient.ts

```diff
@@ -1,36 +1,54 @@
 import { PoapMomentsApi, CompassProvider } from '@poap-xyz/providers';
 import { PaginatedResult, nextCursor } from '@poap-xyz/utils';
-import { CreateMomentInput, FetchMomentsInput } from './types';
-import { Moment } from './domain/Moment';
+import { Moment } from '../domain/Moment';
 import {
   createBetweenFilter,
   createFilter,
   createInFilter,
   creatEqFilter,
   filterUndefinedProperties,
-} from './queries/utils';
+} from '../queries/utils';
 import {
   MomentResponse,
   MomentsQueryResponse,
   PAGINATED_MOMENTS_QUERY,
-} from './queries';
+} from '../queries';
+import { CreateMomentInput } from './dtos/create/CreateInput';
+import { CreateSteps } from './dtos/create/CreateSteps';
+import { FetchMomentsInput } from './dtos/fetch/FetchMomentsInput';
 
 export class MomentsClient {
   constructor(
-    private PoapMomentsApi: PoapMomentsApi,
+    private poapMomentsApi: PoapMomentsApi,
     private CompassProvider: CompassProvider,
   ) {}
 
-  async createMoment(input: CreateMomentInput): Promise<Moment> {
-    const { url, key } = await this.PoapMomentsApi.getSignedUrl();
-    await this.PoapMomentsApi.uploadFile(input.file, url, input.fileType);
-    await this.PoapMomentsApi.waitForMediaProcessing(key, input.timeOut);
-    const response = await this.PoapMomentsApi.createMoment({
+  public async createMoment(input: CreateMomentInput): Promise<Moment> {
+    const defaultOnStepUpdate = (): void => {
+      //do nothing
+    };
+    const onStepUpdate = input.onStepUpdate || defaultOnStepUpdate;
```

Comment:
If the given function returns a promise it must be called with async to make sense, isn't?

Reply:
Yes, it could be async or sync, I don't really care how the user implement it because I don't wait for it

Reply:
So when the async function is executed, how it will know that the state is still in that step or have already moved forward? If you don't wait for it, it might be that the later state executes before than a previous state.